Converted back-end to use connect.js, fixes #98

spratt opened the pull request:
YAY WE USE CONNECT NOW WE'RE COOL AND HIP

But seriously, it got rid of a lot of code.

spratt committed:
Added copyright header to server index

spratt committed:
Updated package.json

spratt committed:
Converted back-end to use connect.js

spratt committed:
Fixed most jshint errors

spratt committed:
Added newline to end of dev_mode.js

spratt committed:
Closed configuration comment box in index.js

spratt committed:
Added a request timeout for scaling

spratt committed:
Added log rotation to start script

spratt committed:
start script cleanup

spratt committed:
No longer using formidable directly

We're still using it, just through connect now.

spratt committed:
removed formidable from package.json

spratt commented:
@cdelahousse I'm trying to test the bookmarklet, and I'm finding its build instructions to be impenetrable.  It gives me 3 make targets and doesn't really help me figure out which one (or combination of several?) I need to do.  The buildvendor, for example, suggests that it's important to do this before any vendors can be used?  Do I need to run that before build?  Or does that get run by build?  Does install call build?  Why is git giving me an error if I run make build then make install?  Am I not supposed to do that?  Why isn't the script doing any kind of check?  I'm scared and confused.

Also, how do I use the bookmarklet?  This should really be integrated into our user interface in such a way that users can actually use it.

spratt commented:
Note that this
- closes #92 since it gets rid of the node-static dependency
- closes #99 since it only loads dev_mode (including the static server code) if running in dev mode
- closes #100 same as previous (proxy code is included in dev_mode)

spratt committed:
Fixed most jshint issues

psimonyi commented on bin/start_dev.sh:
```diff
@@ -58,9 +60,16 @@ if [ ! -f $DB_INFO_FILE ]; then
-    1>> $ROOT_DIR/$LOG_FILE 2>> $ROOT_DIR/$LOG_FILE &
+NODE_ENV=development $NODE_EXE $INDEX_SCRIPT &> $ROOT_DIR/$LOG_FILE &
 echo $! > $ROOT_DIR/$PID_FILE
+date +%s > $ROOT_DIR/$TIMESTAMP_FILE
```
`date --iso-8601=seconds` would give more readable filenames.

spratt replied:
That code does not work on my box.

spratt replied:
Would you prefer something like: `date "+%Y/%m/%d/%H:%M"`?

spratt replied:
I guess forward slashes are a dumb idea, so maybe `date "+%Y.%m.%d.%H.%M"`?

psimonyi replied:
On Sat, 2013-03-23 at 23:08 -0700, Simon David Pratt wrote:

> I guess forward slashes are a dumb idea, so maybe date "+%Y.%m.%d.%H.%M"?

Yeah, only ISO 8601 style is probably best:
+'%Y-%m-%dT%H:%M:%S%z'

spratt replied:
Colons are a bad idea, they're the file delimiter on mac.

spratt replied:
*Path delimiter.

psimonyi replied:
They're path delimiters on *nix, but still okay in path names. However,
it's fine to have no delimiters:
+'%Y-%m-%dT%H%M%S%Z'

spratt replied:
By path delimiter, I mean mac represents paths as usr:bin:foo

psimonyi replied:
How do they use all the BSD and GNU stuff that assumes : is a path
delimiter (as in PATH=/bin:/usr/bin:$HOME/bin)? Anyway, the standard
allows you to omit the separators.

spratt replied:
Yeah, they do some weird conversion magic so it looks like paths are separated by /.

spratt committed:
Edited timestamps to be more readable

psimonyi commented on src/server/index.js:
```diff
@@ -1,71 +1,80 @@
+  .use(connect.logger(devMode?'dev':'short'))
+  .use(connect.timeout(requestTimeout))
+  .use(function(request,response,next) {
+    // Preflight OPTIONS requests -- CORS
```
I think this means we'll allow cross-origin requests for everything, when we used to allow them only for `/do/newcode`, right?

spratt replied:
Yup, totally forgot.

psimonyi commented on src/server/package.json:
```diff
@@ -1,43 +1,38 @@
-  }
-  , "license": "ISC License"
-  , "engine" : ">=0.7.9-pre"
+    "url":  "https://bitbucket.org/sdp/codereviewservermaster.git"
```
This doesn't look current...

spratt replied:
Yeah, this is our old repo.

psimonyi commented on src/server/package.json:
```diff
@@ -1,43 +1,38 @@
+    "url":  "https://bitbucket.org/sdp/codereviewservermaster.git"
+  },
+  "license": "ISC License",
+  "engine" : ">=0.7.9-pre"
```
Is this accurate?

spratt replied:
Not a clue.  Want me to update it to the current version I'm using?

psimonyi replied:
I guess it should be the minimum version any of us test with. (For me,
it's v0.10.1.) So sure, set it to whatever you're using.

spratt replied:
Am also using 10.1.

psimonyi commented on src/server/languageList.js:
I like the diffstats on this file!

spratt committed:
Now only allowing CORS by a whitelist

spratt committed:
Updated git repo url in package.json

spratt commented on src/server/languageList.js:
Yeah, we weren't using it anyway....I think I just forgot to delete it a while back.

psimonyi commented on bin/start_dev.sh:
```diff
@@ -58,9 +61,16 @@ if [ ! -f $DB_INFO_FILE ]; then
 fi
 
 ######################################################################
+# Rotate logs
```
Nobody's developing this on Windows, right? So how about putting the log file in the right place and keeping a symlink to the latest one? As in, `node index.js > var/log/server-$TIME.log; ln -s -f var/log/server-$TIME.log var/server.log` (approximately).

spratt replied:
If they're running the shell script, they would have to be using cygwin which I'm pretty sure supports symbolic links.  It feels a little weird to do this, but I think it's more elegant than saving a timestamp in an extra file.

spratt committed:
Now var/server.log is a symbolic link to the latest log

Also, the start script checks for outdated npm packages and only
prints the message that it's resolving npm dependencies if any
packages are missing or outdated.

	modified:   bin/start_dev.sh

spratt committed:
Fixed bug where server pid wasn't stored properly

spratt committed:
Fixed typo in start_dev.sh

spratt committed:
Replaced our CORS code with middleware package

psimonyi commented on .gitignore:
```diff
@@ -15,3 +15,4 @@ npm-debug.log
 *.db
 *.DS_Store
 *~
+*.time
```
This is no longer needed.

spratt replied:
Fixed in    ba5c966

spratt committed:
Removed unneeded *.time in .gitignore

psimonyi commented on bin/start_dev.sh:
```diff
@@ -35,17 +38,15 @@ fi
-echo "Installing needed packages"
-npm install
+OUTDATED=`npm outdated 2> /dev/null`
+if [ ${#OUTDATED} -ne 0 ]; then
```
Instead of comparing the string length to 0, you should just check whether the string is nonempty.

```
if [ "$(npm outdated 2> /dev/null)" ]; then
```

is most concise. (You could use backticks instead of `$()`; I just think it looks odd inside quotation marks.)
Other options include `if [ "$OUTDATED" ]` and `if [ -n "$OUTDATED" ]` if you'd rather keep the variable.

spratt replied:
Checking the length is much clearer than what seems to be using a string as a boolean.

psimonyi replied:
Use the `-n` operator: `if [ -n "$(npm outdated 2> /dev/null)" ]; then`

And without `-n`, you're not using a string as a boolean; you're passing a single argument to the `test` builtin.

spratt replied:
Agreed, so it's proper syntax for test but it's not syntax whose meaning is immediately obvious.  Also, I'm getting a substitution error when I try to evaluate that line within the brackets.

spratt replied:
It looks like ${command} is not equivalent to `command` on my system.

psimonyi replied:
On Sun, 2013-03-24 at 19:57 -0700, Simon David Pratt wrote:

> It looks like ${command} is not equivalent to command on my system.
> Parentheses, not braces. Round brackets, if you like.

spratt replied:
Oh herp.  Well, I fixed with backticks.  Good enough?

psimonyi replied:
On Sun, 2013-03-24 at 19:57 -0700, Simon David Pratt wrote:

> It looks like ${command} is not equivalent to command on my system.
> In Bash, $(command) is the same as `command`.

spratt replied:
I understand that from your previous comment.  I am saying that the pull request now includes your fix with backticks inside of quotes instead of the parens.  Want me to go back and replace the backticks with parens?

psimonyi replied:
Nah, backticks are fine. We're using them all over anyway.

spratt committed:
Now checking if string is null rather than length is 0

spratt commented:
Oh man, I'm looking forward to pull request I have waiting to submit after this one.

psimonyi commented:
On Sun, 2013-03-24 at 21:32 -0700, Simon David Pratt wrote:

> Oh man, I'm looking forward to pull request I have waiting to submit
> after this one.
> You can't have two pull requests at the same time?

spratt commented:
I can, but the other one is a fork from these changes.  The diff would include all of these changes.

psimonyi commented on src/server/index.js:
```diff
@@ -1,71 +1,70 @@
+* Start the server                                                            *
+******************************************************************************/
+var app = connect()
+  .use(connect.logger(devMode?'dev':'short'))
```
I'd prefer spaces around the `?` and `:` (but what I'd like most is to resolve the "Proposals" in the style guide).

spratt replied:
I like all the proposals.  It's been the de facto style guide for some time now.  Let's consider it resolved.

psimonyi replied:
On Sun, 2013-03-24 at 22:15 -0700, Simon David Pratt wrote:

> I like all the proposals. It's been the de facto style guide for some
> time now. Let's consider it resolved.
> Okay, well, there are a lot of places here with different spacing: you
> mostly seem to put no space between a control structure keyword and the
> paren starting its head, for example.

spratt replied:
I figured the jshint would catch those errors.  I guess I'll go back and check everything against the style guide.

spratt committed:
Fixed style error in index.js

spratt committed:
Updated to match code style guide
